### PR TITLE
kerosene: Add product method to calculate cartesian product of arrays

### PR DESCRIPTION
**Pull request by nhardy**

### CODE REVIEW

File: packages/kerosene/src/array/product.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * Returns the cartesian product of the source arrays
+ *
+ * Usage:
+ * ```typescript
+ * product([1, 2], [3, 4]); // [[1, 3], [1, 4], [2, 3], [2, 4]]
+ * ```
+ *
+ * @param sources
+ */
+export default function product<T>(...sources: T[][]): T[][] {
```
Comment (brycehanscomb) [inline packages/kerosene/src/array/product.ts]: Can you make `T` optional? We wouldn't want to enforce typings on such a generic fn

File: testSetup.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import "core-js/features/array/flat-map";
```
Comment (brycehanscomb) [inline testSetup.ts]: This makes our modules no longer `pure`, right?

Comment (brycehanscomb) [reply testSetup.ts]: re: #24

File: packages/kerosene/src/array/product.ts (hunk 1 above)
Comment (nhardy) [reply packages/kerosene/src/array/product.ts]: `T` will be inferred from the type of the input `sources`, so providing this isn't required for a consumer.

File: testSetup.ts (hunk 2 above)
Comment (nhardy) [reply testSetup.ts]: This is just for test setup and isn't included for the actual build. But you do raise an important point. I know that we are using ES2015+ `Array.prototype` methods like `.every` and `.includes`, but perhaps we should be more specific about BYO polyfills?

Comment (brycehanscomb) [reply testSetup.ts]: good point. Can you raise a quick Issue so we can talk about it sometime

**Pull request merged**